proposal/public-compiler-api.md

```diff
@@ -0,0 +1,172 @@
+# Public Compiler API
+
+## Background
+
+> This section is non-normative.
+
+Sass compiler implementations tend to assume projects will have few entrypoint Sass files
+such that running the compiler against each without a cache is "fast enough".
+However, for projects with substantially more entrypoints, ones that depend
+on a bundler (like webpack) to do final concatenation, or using [css-modules][], there can be a significant cost to reparsing and processing unchanged files across the project.
+
+In common case of projects with many entrypoints, each entry likely imports from a
+large set of shared configuration, funtions and mixins. These dependencies often change infrequently
+and a significant amount of compile time is spent compiling the same files in the
+context of a new entrypoint file.
+
+In addition, without an ability to reuse previous work, certain project patterns with Sass's
+module system are also impossible. Consider this case using webpack:
```

Comment:
I also want to maintain the explicit nature of the new module system! I think whether or not this use-case fits that depends on the context and framing. If you consider each Sass entrypoint a separate "program", sharing module state like this doesn't make sense and is confusing. 

In the context of bundlers though, you aren't usually writing entry-points as distinct programs but as branches in a single "App" program. `button.scss` is functionally a "partial" in this example, e.g. I don't intend it to be it's own css file and know the bundler will combine it into the final asset. 

Of course one could write these as actual partials and have a single or limited number of hand tuned entry files  but that's not organizationally great when you have hundreds of partials in the app. It'd be better (I think) if to let the bundler app bundler create the ideal top level entrypoints based on the overall dependency graph. Ideally this enables smaller initial css downloads, and optimized outputs for above/below the fold.

If you think the above distinction makes sense, I'd argue that the proposed feature isn't all that useful for the first case. Sass is already fast enough for cases where you have a limited number of top level entrypoints. I'd suggest that the need for this feature really only arises when writing apps in the context of a bundler assuming the second framing. e.g. most every import of a style file is a partial. 

My suggestion is that the example outlined is fundamentally the same as this: 

```styles.scss
@use 'bootstrap' with (
  $btn-color: red
)
@use 'button.scss'
```
_button.scss
```
@use 'bootstrap/buttons' as buttons

.btn {...}
```

Except that to trace the interactions you need to jump through JS and Sass files, but it is no less explicit (albeit a much larger) graph. 

OK Philosophizing aside, I do think that this approach offers actual benefits:

### It removes a the need for large top level "god" files that reexport all variables, functions and mixins

Even with the new module system you still end up with a giant "theme.scss" files that re-export everything since you need to do `@use .. with` overrides for every entrypoint, making it less useful to be able to import just the bits you need from submodules.

### It's easily to structure libraries that can be configured without duplicating CSS in the output

As with using `@import` providing customization of libraries requires manually split out the files that produce css vs contain variables and other members. I think it'd be awesome if one could just `@use 'bootstrap' with (...)` at the app root to configure and include the bootstrap css and then be able to import the mixins and functions I need in the app without needing to manually split the `bootstrap.scss` into configuration with override file and css producing partials. (This could also be addressed by having something like Less's `@import (reference)` maybe?)

I have more thoughts and am **rambling** but it's Saturday and I should step away for now, sorry for the wall of text, I'm glad to have a chance to try and articulate why I think this is important for Sass in the context of all the crazy tooling we have in the web dev space now.

Reply:
There are a lot of different thoughts in this comment and I'm not going to try to address all of them, but at a high level, I don't think there's a compelling case for adding implicit dependencies between entrypoints. Even your example of configuring `bootstrap` and using `button` in the same file is something I would flag as unclear as a code reviewer, so doubling down on that pattern is not something I want to do.

I'll try to address the two main benefits you bring up:

> It removes a the need for large top level "god" files that reexport all variables, functions and mixins

You don't necessarily need to re-export anything, you just need to be explicit about your dependencies. In this case, you're depending on your app's local configuration, so you can write:

```scss
// button.scss
@use 'myapp/configuration'; // configures bootstrap
@use 'bootstrap/buttons';
```

This is both simple and explicit.

> It's easily to structure libraries that can be configured without duplicating CSS in the output

It's pretty easy to write a library that doesn't duplicate CSS for simple cases: just put all the CSS generation in mixins that downstream clients can choose to invoke or not. Beyond that, doing this at scale requires a proper code-splitting solution which is an explicit non-goal of the Sass module system.

Reply:
> your example of configuring bootstrap and using button in the same file is something I would flag as unclear as a code reviewer, so doubling down on that pattern is not something I want to do.

Interesting, If you don't mind how you write or structure that example then? It's seems like consumers of configured modules are always going to have no knowledge of whether they were configured or not?

> @use 'myapp/configuration'; // configures bootstrap
> @use 'bootstrap/buttons';

Definitely, you don't need to reexport everything but the cost of not doing that, is needing to remember to configure everything before accessing it in many, many files, which is error prone.

>  Beyond that, doing this at scale requires a proper code-splitting solution which is an explicit non-goal of the Sass module system.

I think we agree on that, but without an API to control the module configuration cache it's not possible for other tools (like bundlers) to implement such a system.

>  just put all the CSS generation in mixins that downstream clients can choose to invoke or not.

That is a cool approach I had not thought of! I've mostly only seen projects structured like bootstrap, which definitely does not follow that pattern


Use-case aside, I'm not actually sure how to move forward here if module configuration needs to be "reset" between entry points. It seems like there is not a lot of reusable work unless there is an easy way cache based on the configuration? I will admit that i do not know how module configuration state is managed tho, it seems like it'd be harder to know whether `with` was invoked too many times in the `compileMany` case?

Reply:
> Interesting, If you don't mind how you write or structure that example then? It's seems like consumers of configured modules are always going to have no knowledge of whether they were configured or not?

I'd structure it like the example I gave above: explicitly using a configuration file in each entrypoint.

> Definitely, you don't need to reexport everything but the cost of not doing that, is needing to remember to configure everything before accessing it in many, many files, which is error prone.

When you're dealing with side-effects (which `@use ... with` is), they're *always* going to be error-prone. I don't think it's less error-prone for being implicit; you're just shuffling around where the errors are likely to happen.

For example, in your proposal as it stands, the behavior of a given entrypoint is highly dependent on the order in which entrypoints are compiled (which is left up to the caller). If a caller decides to compile `button.scss` before `styles.scss` then the configuration won't be available in `button.scss` for reasons that may be unclear to the author. If a caller compiles stylesheets in parallel, there may be a race condition to determine whether or not the configuration is available.

It gets even worse if some entrypoint changes the state of a shared module in some way other than `@use ... with`. If `button.scss` sets `bootstrap.$base-color: #000;` then entrypoints that run after `button.scss` will use a different color scheme than any that run before it.

> I think we agree on that, but without an API to control the module configuration cache it's not possible for other tools (like bundlers) to implement such a system.

I don't follow this logic.

> Use-case aside, I'm not actually sure how to move forward here if module configuration needs to be "reset" between entry points. It seems like there is not a lot of reusable work unless there is an easy way cache based on the configuration? I will admit that i do not know how module configuration state is managed tho, it seems like it'd be harder to know whether `with` was invoked too many times in the `compileMany` case?

I think the question of what work the compiler can re-use can be separated to a degree from the question of which guarantees the compiler API offers to its users. Defining the latter is the primary goal of this design, which will then allow the compiler to freely optimize within the constraints of those guarantees.

There are certainly some valuable optimizations we could do here, such as re-using the ASTs of files that haven't changed between compilations and even re-using the loaded modules between compilations under certain constraints.

Reply:
> I think the question of what work the compiler can re-use can be separated to a degree from the question of which guarantees the compiler API offers to its users.

Sorry i'm not sure I understand this distinction. Isn't the question of what is resusable dependent on what requires compiling? The latter doesn't seem knowable unless the question of whether a module needs configuration applied. e,g. the result of module A is different depending on how it was imported, and seems like those results need to be cached separately, Or hand wave past it as "reevaluate as necessary" or something?

> For example, in your proposal as it stands, the behavior of a given entrypoint is highly dependent on the order in which entrypoints are compiled (which is left up to the caller)

Yes i'd consider that a feature not a bug. It'd absolutely error prone when used incorrectly, tho that's often the case with low level APIs i think. TBF i've been envisoning this API as hook for more complicated use cases and specifically intended to be used by other tools, not directly by users to compile Sass. A core assumption of mine is that is those who would benefit from this API are not using the Sass compiler directly anyway. I'm not doing a good job of championing or explaining the potential benefit here in the context of other tools, but that is were I'm coming from, Sass as part of a larger application graph.